File: src/main/java/com/dochiri/hexagonal/infrastructure/user/mapper/UserMapper.java
package com.dochiri.hexagonal.infrastructure.user.mapper;

import com.dochiri.hexagonal.domain.user.User;
import com.dochiri.hexagonal.domain.user.vo.Email;
import com.dochiri.hexagonal.domain.user.vo.UserId;
import com.dochiri.hexagonal.infrastructure.user.entity.UserEntity;
import org.springframework.stereotype.Component;

@Component
public class UserMapper {

    public UserEntity toEntity(User domain) {
        return new UserEntity(
                domain.getId().value(),
                domain.getEmail().value(),
                domain.getPasswordHash(),
                domain.getName(),
                domain.getStatus().name(),
                domain.getRole().name(),
                domain.getLastLoginAt()
        );
    }

    public User toDomain(UserEntity entity) {
        return User.of(
                UserId.from(entity.getPublicId()),
                Email.from(entity.getEmail()),
                entity.getPasswordHash(),
                entity.getName(),
                UserStatus.valueOf(entity.getStatus()),
                UserRole.valueOf(entity.getRole()),
                entity.getLastLoginAt()
        );
    }

    public void applyFullUpdate(User domain, UserEntity entity) {
        entity.updateFromDomain(
                domain.getEmail().value(),
                domain.getPasswordHash(),
                domain.getName(),
                domain.getStatus().name(),
                domain.getRole().name(),
                domain.getLastLoginAt()
        );
    }

}

File: src/main/kotlin/com/dochiri/hexagonal/infrastructure/user/mapper/UserMapper.kt
package com.dochiri.hexagonal.infrastructure.user.mapper

import com.dochiri.hexagonal.domain.user.User
import com.dochiri.hexagonal.domain.user.UserRole
import com.dochiri.hexagonal.domain.user.UserStatus
import com.dochiri.hexagonal.domain.user.vo.Email
import com.dochiri.hexagonal.domain.user.vo.UserId
import com.dochiri.hexagonal.infrastructure.user.entity.UserEntity

fun User.toEntity(): UserEntity =
    UserEntity(
        publicId = id.value,
        email = email.value,
        passwordHash = passwordHash,
        name = name,
        status = status.name,
        role = role.name,
        lastLoginAt = lastLoginAt
    )

fun UserEntity.toDomain(): User =
    User.reconstitute(
        UserId.from(publicId),
        Email.from(email),
        passwordHash,
        name,
        UserStatus.valueOf(status),
        UserRole.valueOf(role),
        lastLoginAt
    )

fun UserEntity.applyFullUpdate(domain: User) {
    updateFromDomain(
        email = domain.email.value,
        passwordHash = domain.passwordHash,
        name = domain.name,
        status = domain.status.name,
        role = domain.role.name,
        lastLoginAt = domain.lastLoginAt
    )
}
